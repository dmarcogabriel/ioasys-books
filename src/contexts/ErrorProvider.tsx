import React, {createContext} from 'react';
import {UNAUTHORIZED_STATUS} from '../constants';
import Toast from 'react-native-toast-message';

type ErrorData = {
  message: string;
};

export interface ApiError {
  errors: ErrorData;
}

interface ErrorContextValue {
  handleError: (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ) => Promise<void>;
}

export const ErrorContext = createContext<ErrorContextValue>({
  handleError: async () => {},
});

interface ErrorProviderProps {
  children: React.ReactNode;
}

export const ErrorProvider = ({children}: ErrorProviderProps): JSX.Element => {
  const handleError = async (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ): Promise<void> => {
    let errorMessage: string = '';
    if (status === UNAUTHORIZED_STATUS) {
      await logoutCallback();
      errorMessage = 'Sess√£o expirada';
    } else {
      Toast.show({
        position: 'bottom',
        text1: errorMessage || error.errors.message,
        type: 'error',
      });
    }
  };

  return (
    <ErrorContext.Provider value={{handleError}}>
      {children}
    </ErrorContext.Provider>
  );
};
