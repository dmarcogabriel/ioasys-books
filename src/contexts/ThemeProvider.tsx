import React from 'react';
import {ThemeProvider as SCThemeProvider} from 'styled-components/native';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export const ThemeProvider = ({children}: ThemeProviderProps): JSX.Element => {
  const theme = {
    colors: {
      primary: '#B22E6F',
      dark: '#333333',
      light: '#ffffff',
      grey: '#999999',
    },
  };

  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};
