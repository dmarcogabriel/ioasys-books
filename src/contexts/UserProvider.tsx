import React, {createContext, useState, useEffect, useCallback} from 'react';
import {User} from '../@types/user.interface';
import {storeData, removeData, getData} from '../utils/storage';
import {AUTH_TOKEN_STORAGE_KEY} from '../constants';

interface UserContextValue {
  user: User | null;
  login: (user: User, token: string) => Promise<void>;
  logout: () => Promise<void>;
  token: string | null;
}

export const UserContext = createContext<UserContextValue>({
  user: null,
  token: null,
  login: async () => {},
  logout: async () => {},
});

interface UserProviderProps {
  children: React.ReactNode;
}

export const UserProvider = ({children}: UserProviderProps): JSX.Element => {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const login = async (data: User, authToken: string): Promise<void> => {
    setUser(data);
    setToken(authToken);
    await storeData(AUTH_TOKEN_STORAGE_KEY, authToken);
  };

  const logout = async () => {
    await removeData(AUTH_TOKEN_STORAGE_KEY);
    setUser(null);
    setToken(null);
  };

  const loadToken = useCallback(async () => {
    setIsLoading(true);
    try {
      const authToken = await getData<string>(AUTH_TOKEN_STORAGE_KEY);
      if (authToken) {
        setToken(authToken);
      }
    } catch (error) {
      setIsLoading(false);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    loadToken();
  }, [loadToken]);

  return (
    <UserContext.Provider value={{user, token, login, logout}}>
      {!isLoading && children}
    </UserContext.Provider>
  );
};
