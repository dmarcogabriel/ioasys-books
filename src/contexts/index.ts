export * from './UserProvider';
export * from './ThemeProvider';
export * from './ErrorProvider';
