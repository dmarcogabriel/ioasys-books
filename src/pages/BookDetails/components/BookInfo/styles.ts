import styled, {css} from 'styled-components/native';
import {Text} from 'react-native-elements';
import {ThemeProps} from '../../../../@types/theme.interface';

export const BookInfoContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
`;

interface BookInfoTextProps {
  isBold: boolean;
}

export const BookInfoText = styled(Text)`
  ${({theme, isBold}: ThemeProps & BookInfoTextProps) => css`
    color: ${isBold ? theme.colors.dark : theme.colors.grey};
    font-weight: ${isBold ? 'bold' : 'normal'};
  `}
`;
