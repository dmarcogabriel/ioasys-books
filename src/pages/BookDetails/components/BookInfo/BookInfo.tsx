import React from 'react';
import {BookInfoContainer, BookInfoText} from './styles';

export interface BookInfoProps {
  name: string;
  value: string | number;
}

export const BookInfo = ({name, value}: BookInfoProps): JSX.Element => (
  <BookInfoContainer>
    <BookInfoText testID="bookInfoName" isBold>
      {name}
    </BookInfoText>
    <BookInfoText testID="bookInfoValue" isBold={false}>
      {value}
    </BookInfoText>
  </BookInfoContainer>
);
