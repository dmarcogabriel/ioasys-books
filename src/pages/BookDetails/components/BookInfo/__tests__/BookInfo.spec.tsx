import React from 'react';
import {render} from '@testing-library/react-native';
import {BookInfo, BookInfoProps} from '../BookInfo';
import {ThemeProvider} from '../../../../../contexts';
import renderer from 'react-test-renderer';

const ComponentWrapper = (props: BookInfoProps) => (
  <ThemeProvider>
    <BookInfo {...props} />
  </ThemeProvider>
);

describe('Books/components/BookCard', () => {
  it('should render correctly', () => {
    const {getByTestId} = render(
      <ComponentWrapper name="Test" value="Is Test" />,
    );
    const bookInfoName = getByTestId('bookInfoName');
    const bookInfoValue = getByTestId('bookInfoValue');

    expect(bookInfoName).toHaveProp('isBold', true);
    expect(bookInfoValue).toHaveProp('isBold', false);
    expect(bookInfoName.children[0]).toBe('Test');
    expect(bookInfoValue.children[0]).toBe('Is Test');
  });

  it('should match snapshot', () => {
    expect(
      renderer
        .create(<ComponentWrapper name="Test" value="Is Test" />)
        .toJSON(),
    ).toMatchSnapshot();
  });
});
