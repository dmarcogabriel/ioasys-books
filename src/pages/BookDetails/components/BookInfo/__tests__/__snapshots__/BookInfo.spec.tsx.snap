// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books/components/BookCard should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
    ]
  }
>
  <Text
    isBold={true}
    style={
      Array [
        Object {
          "color": "#333333",
          "fontWeight": "bold",
        },
      ]
    }
    testID="bookInfoName"
  >
    Test
  </Text>
  <Text
    isBold={false}
    style={
      Array [
        Object {
          "color": "#999999",
          "fontWeight": "normal",
        },
      ]
    }
    testID="bookInfoValue"
  >
    Is Test
  </Text>
</View>
`;
